### PR TITLE
chore: Parse Datadog Log agent payloads faster

### LINKED ISSUE #8286: `datadog_logs` sink should eschew `decode_body`
In [this comment](https://github.com/timberio/vector/issues/8263#issuecomment-879417074) we demonstrate that there's a performance benefit to eschewing `decode_body` in the Datadog logs sinks, worth approximately 20Mb/s in the setup described in the linked issue. The linked issue includes a diff that makes the conversion though it's not clear yet if the `LogMsg` is actually representative of the datadog agent log payload. Tests do pass.

### PR DESCRIPTION
**Pull request by blt**
This commit adjusts the way the datadog agent log payloads are parsed, avoiding
the generic `decode_body` and relying on our knowledge of the payload's
structure to avoid the branch misses etc etc inherent in that function. We gain
20Mb/s of parse speed per investigation done in #8263. This commit passes
tests (locally) but I don't have documentation on hand to assert that this
payload is, in fact, what the datadog agent is meant to send.

Resolves #8286

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
It'd be good to get confirmation from the dd-agent team on the message fields. At least on https://docs.datadoghq.com/api/latest/logs/#send-logs I see more fields. We should test this with the `datadog_logs` Vector sink too.

Otherwise, this makes sense to me and is a great improvement in performance!

**Comment by blt** [comment]
> It'd be good to get confirmation from the dd-agent team on the message fields. At least on https://docs.datadoghq.com/api/latest/logs/#send-logs I see more fields. We should test this with the `datadog_logs` Vector sink too.
> 
> Otherwise, this makes sense to me and is a great improvement in performance!

Agreed. I don't want this merged until we have confirmation from folks that know that these fields are sufficient. FWIW I haven't reached out to the team yet, being content to wait for either @prognant or @vladimir-dd to take a look. 

Regarding the docs you linked, I wasn't sure if that is the external payload or what.

**Comment by prognant** [comment]
TL;DR: the Datadog logs json schema is a flat schema with [7 fields](https://github.com/DataDog/datadog-agent/blob/main/pkg/logs/processor/json.go#L23-L32).

A typical logs sent by the Agent would look like:
``` 
{
    "message": "[Mon Mar  1 14:29:21 CET 2001] Hello Dave !",
    "status": "notice",
    "timestamp": 1614605362706,
    "hostname": "hal",
    "service": "hal9000",
    "ddsource": "obc",
    "ddtags": "env:space"
}
```
On top of documented fields (https://docs.datadoghq.com/api/latest/logs/#send-logs) this adds the `status` fields.
The agent serialises logs [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/logs/processor/json.go#L23-L49) so there cannot be more than those 7 fields (I don't think there is any plan to change that).

There is however a slight variation related to log collection in AWS lambda where some additional runtime context are added to each [log record](https://github.com/DataDog/datadog-agent/blob/main/pkg/logs/processor/json_serverless.go#L23-L72), that being said I don't think we planned to support aggregating that as Lambdas don't run on-prem (Maybe with Oupost, didn't check).

**Comment by prognant** [comment]
There is however a caveat to my previous post, while the Datadog agent will strictly comply with the 7 fields schema, datadog supports random extra json fields, if you do:
```
$ curl -X POST "https://http-intake.logs.datadoghq.com/v1/input" \
 -H "Content-Type: application/json" \
 -H "DD-API-KEY: ${DD_API_KEY}" \
 -d @- << EOF
 {
   "message": "Hello!",
   "foo": "bar",
   "hostname": "prognant-dev-laptop"
 }
EOF
```

The field+value `"foo": "bar"` will be usable on the datadog side more or less like a standard tags, it will appear as an attributes in the log explorer, and query is supported on those attributes:
image

I think this predates only supporting the schema the Agent is using.

**Comment by blt** [comment]
> There is however a caveat to my previous post, while the Datadog Agent will strictly comply with the 7 fields schema, Datadog supports random json fields, if you do:
> 
> ```
> $ curl -X POST "https://http-intake.logs.datadoghq.com/v1/input" \
>  -H "Content-Type: application/json" \
>  -H "DD-API-KEY: ${DD_API_KEY}" \
>  -d @- << EOF
>  {
>    "message": "Hello!",
>    "foo": "bar",
>    "hostname": "prognant-dev-laptop"
>  }
> EOF
> ```
> 
> The field+value `"foo": "bar"` will be usable on the Datadog side more or less like a standard tag, it will appear as an attributes in the log explorer, and queries are supported on those attributes:
> image
> 
> I think this predates only supporting the schema the Agent is using.

Got it. This seems very relevant to the sink side of things where I was intending to pull the same specialization trick. I will keep that in mind.

File: src/sources/datadog/logs.rs [hunk 1]
```diff
@@ -78,10 +78,16 @@ struct DatadogLogsSource {
...
 struct LogMsg {
-    pub message: String,
-    pub timestamp: i64,
+    pub message: Bytes,
+    pub status: Option<Bytes>,
+    pub timestamp: Option<i64>,
+    pub hostname: Option<Bytes>,
+    pub service: Option<Bytes>,
+    pub source: Option<Bytes>,
+    pub tags: Option<Bytes>,
```
Comment (jszwedko) [inline src/sources/datadog/logs.rs:90]: Does this still need a `HashMap` to pick up the remaining arbitrary key/values?

Comment (blt) [reply src/sources/datadog/logs.rs:90]: My understanding from the agent code is that there wouldn't be further fields.

Comment (blt) [reply src/sources/datadog/logs.rs:90]: Ah, I replied to a pre-edit. Well, that's a good question. Does this source support arbitrary clients that are allowed to send arbitrarily keyed messages or is it specifically to catch log payloads from the datadog agent?

Comment (blt) [reply src/sources/datadog/logs.rs:90]: If we want the performance story here to be as good as possible we need to be specific about the keys that are inbound so we can map them to a struct without indirection into a collection but I'm also sympathetic to the idea that some users will end up retargetting their local vector install instead of the intake API, in which case this source needs to be as lax as that API about what inputs it accepts.

Comment (jszwedko) [reply src/sources/datadog/logs.rs:90]: Sorry, that edit could have probably been another comment. And yes, that is a good question 🙂

Historically, we've tried to keep sources/sinks compatible such that I could, for example, use the `datadog_logs` sink to push to a vector with a `datadog_logs` source. That alone seems like it'd require handling of the dynamic fields.

Having the structured fields you have here, along with a catch-all, might still be enough to see the performance benefit with logs coming from the datadog agent, depending on how serde handles the deserialization. However, if there is still a significant performance penalty, we could consider making it something you have to opt into.

Comment (blt) [reply src/sources/datadog/logs.rs:90]: We do seem to suggest in the source documentation that it's a datadog agent doing the sending, so if we mean to be more expansive here it's worth documenting what the correct payload form is.

Comment (jszwedko) [reply src/sources/datadog/logs.rs:90]: Chatted with Ben about this and we think we could make the source specific to receiving from the agent with some additional notes:

* We rename this `datadog_logs` source to `datadog_agent`. This `datadog_agent` source will be the primary source for receiving all data from the agent (logs, metrics, traces). To start, we'd only support logs. Metrics is expected to come this quarter. The rename would also help make it clear that this is a "breaking change"; I think we could avoid adding the alias and have people explicitly opt over into it.
* We get a "contract" in-place with the Datadog agent team so that they notify us if the agent is going to modify the way it sends data to allow us to get a release out ahead of their changes to accommodate them.
* In the future, we may reintroduce a `datadog_logs` source that is intended to mimic the Datadog Logs ingest endpoint and would be fully compatible with the `dattadog_logs` sink.

Comment (blt) [reply src/sources/datadog/logs.rs:90]: @jszwedko perfect. I think my appetite here is to get this PR merged and we follow-up with a rename / re-document issue since I want to be sure to isolate changes as much as possible for measurement purposes.

**Review by bruceg — COMMENTED** [review]
Looks reasonable given the fixed schema from the agent. I wonder if it could be even more fixed.

File: src/sources/datadog/logs.rs [hunk 2]
```diff
@@ -77,6 +78,18 @@ struct DatadogLogsSource {
     store_api_key: bool,
 }
 
+// https://github.com/DataDog/datadog-agent/blob/a33248c2bc125920a9577af1e16f12298875a4ad/pkg/logs/processor/json.go#L23-L49
+#[derive(Deserialize)]
+struct LogMsg {
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sources/datadog/logs.rs:83]: Shoot, yes it should.

File: src/sources/datadog/logs.rs [hunk 3]
```diff
@@ -77,6 +78,18 @@ struct DatadogLogsSource {
...
+    pub message: Bytes,
+    pub status: Option<Bytes>,
+    pub timestamp: Option<i64>,
+    pub hostname: Option<Bytes>,
+    pub service: Option<Bytes>,
+    pub source: Option<Bytes>,
+    pub tags: Option<Bytes>,
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sources/datadog/logs.rs:90]: Ah, that's a good question. Based on the _tests_ I assumed that the agent would sometimes not send every field but I think we can drop all of these `Options` which would be great! The code is here: https://github.com/DataDog/datadog-agent/blob/a33248c2bc125920a9577af1e16f12298875a4ad/pkg/logs/processor/json.go#L23-L49

None of these have an `omitempty` so unless there's some Go subtly I'm unaware of we should be expecting every field in every message.

[the remark this answers is not in the document]
Comment (blt) [reply src/sources/datadog/logs.rs:90]: @prognant is this understanding correct?

File: src/sources/datadog/logs.rs (hunk 1 above)
Comment (jszwedko) [reply src/sources/datadog/logs.rs:90]: 👍 as it stands, this merge would be breaking. Would it be ok if I open a PR to move around `datadog_logs` first? I can do it tomorrow.

Comment (blt) [reply src/sources/datadog/logs.rs:90]: Sure. I'm in no hurry and am happy to rebase this small change however.

File: src/sources/datadog/logs.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/sources/datadog/logs.rs]: Alright, I have removed the `Option`s and adjusted how we build the `LogEvent` some and the whole situation is _pretty fast_. Moving the call to `log_schema()` up out of the main loop of the decode function removed a serialization point -- each lazy static is an atomic read, each sync'ed `OnceCell` is too -- and my Thinkstation is breezing along at ~300Mb/s throughput, per the setup discussed in #8263, with vector restricted to 8 cores.

File: src/sources/datadog/logs.rs [hunk 4]
```diff
@@ -77,6 +78,56 @@ struct DatadogLogsSource {
...
+        )
+    })?;
+    let timestamp_key = log_schema().timestamp_key();
+    let source_type_key = log_schema().source_type_key();
+    let now = Utc::now();
+    let iter = messages
+        .into_iter()
+        .map(|msg| {
+            let mut log = LogEvent::default();
+            log.insert_flat(timestamp_key, now);
```
Comment (spencergilbert) [inline src/sources/datadog/logs.rs:112]: What's the interaction if the `timestamp_key` == "timestamp"? Does it conflict with ln116 or are these nested separately?

Comment (blt) [reply src/sources/datadog/logs.rs:112]: Good question. The previous behavior was to overwrite if the log schema's timestamp key was "timestamp". That behavior is preserved here too. 

What _has_ changed, it now occurs to me, is the difference in `insert` versus `insert_flat`. I suppose it's possible that the log schema would be set to insert the timestamp in some subtree which is no longer possible, related to the conversation in #8331.

Comment (spencergilbert) [reply src/sources/datadog/logs.rs:112]: Not to add _more_ complexity but we do have precedence in the `kubernetes_logs` source to record the "ingestion" timestamp and the "event's" timestamp

**Comment by blt** [comment]
Blocked pending #8350.

File: src/sources/datadog/logs.rs (hunk 4 above)
Comment (blt) [reply src/sources/datadog/logs.rs]: @spencergilbert I would support this change in a follow up ticket, probably part of the broader push to tidy up observability?

Comment (spencergilbert) [reply src/sources/datadog/logs.rs]: Sure! Just wanted to surface it, I only recently found out it was a thing too 😄 

I generally can't think of a reason to not use the event's timestamp (unless you want to be able to measure how long it takes for events to go from writing to aggregation 🤷 )

**Review by spencergilbert — APPROVED** [review]
